pkg/controllers/user/authz/podsecuritypolicy/serviceaccount.go

```diff
@@ -27,213 +26,265 @@ func RegisterServiceAccount(context *config.UserContext) {
 }
 
-func (m *serviceAccountManager) sync(key string, obj *v1.ServiceAccount) error {
-	if obj == nil {
-		logrus.Debugf("no service account provided, exiting")
-		return nil
-	}
+func (m *serviceAccountManager) Create(obj *v1.ServiceAccount) (*v1.ServiceAccount, error) {
+	return m.sync(obj)
+}
 
-	namespaceName := obj.Namespace
+func (m *serviceAccountManager) Updated(obj *v1.ServiceAccount) (*v1.ServiceAccount, error) {
+	return m.sync(obj)
+}
 
-	// get PSPT
-	// if no PSPT then get default
-	// if no default then exit
-	namespace, err := m.namespaceLister.Get("", namespaceName)
+func (m *serviceAccountManager) sync(obj *v1.ServiceAccount) (*v1.ServiceAccount, error) {
+	namespace, err := m.namespaceLister.Get("", obj.Namespace)
 	if err != nil {
-		return fmt.Errorf("error getting namespaces: %v", err)
+		return nil, fmt.Errorf("error getting projects: %v", err)
 	}
 
-	annotation := namespace.Annotations[projectIDAnnotation]
-
-	if annotation == "" {
-		// no project is associated with namespace so don't do anything
-		logrus.Debugf("no project is associated with namespace %v", namespaceName)
-		return nil
+	if namespace.Annotations[projectIDAnnotation] == "" {
+		return obj, nil
 	}
 
-	split := strings.SplitN(annotation, ":", 2)
-
-	if len(split) != 2 {
-		return fmt.Errorf("could not parse handler key: %v got len %v", annotation, len(split))
-	}
-
-	clusterName, projectID := split[0], split[1]
-
-	podSecurityPolicyTemplateID, err := getPodSecurityPolicyTemplateID(m.projectLister, m.clusterLister, projectID,
-		clusterName)
+	unfilteredPSPTPBS, err := m.psptpbLister.List("", labels.Everything())
 	if err != nil {
-		return err
+		return nil, fmt.Errorf("error getting psptpbs: %v", err)
 	}
 
-	if podSecurityPolicyTemplateID == "" {
-		// Do nothing
-		logrus.Debugf("no matching pod security policy template")
-		return nil
+	var psptpbs []*v3.PodSecurityPolicyTemplateProjectBinding
+	for _, psptpb := range unfilteredPSPTPBS {
+		if psptpb.DeletionTimestamp == nil {
+			psptpbs = append(psptpbs, psptpb)
+		}
 	}
 
-	policies, err := m.policyLister.List("", labels.Everything())
+	roleBindings, err := m.roleBindingLister.List(obj.Namespace, labels.Everything())
 	if err != nil {
-		return fmt.Errorf("error getting policies: %v", err)
+		return nil, fmt.Errorf("error getting role bindings: %v", err)
 	}
 
-	var policy *v1beta12.PodSecurityPolicy
-
-	for _, candidate := range policies {
-		if candidate.Annotations[podSecurityTemplateParentAnnotation] == podSecurityPolicyTemplateID {
-			policy = candidate
+	onePSPTPBExists := false
+
+	for _, psptpb := range psptpbs {
+		clusterRoleName := getClusterRoleName(psptpb.PodSecurityPolicyTemplateName)
+		if psptpb.TargetProjectName == namespace.Annotations[projectIDAnnotation] {
+			// at least 1 psptpb exists for this project, we need to delete any default role bindings that exist also
+			onePSPTPBExists = true
+
+			for _, roleBinding := range roleBindings {
```

Comment:
RoleBindings for the svc account _will_ exist that are not for PSPs. We cannot assume all bindings for a particualr service account are there because of PSPs.

You should be getting these based on either a label filter or a annotation and index that specificies the binding was made specifically by this handler.

Reply:
I think the delete logic is somewhat backwards in general. I think you need to loop over all of the current service account's PSP bindings, see if it maps to an existing PSPTPB, and delete it if it does not.